### PR TITLE
add pipelineRecord data structure and CRUD

### LINKED ISSUE #160: Add pipeline record data structure and CRUD
Pipeline record is the running record of the pipeline.

### PR DESCRIPTION
**Pull request by zz-bmj**
Fixes #160 
Changes

-  Add pipelineRecord data structure
-  Add pipelineRecord CRUD

### CODE REVIEW

File: pkg/api/types.go [hunk 1]
```diff
@@ -28,6 +28,53 @@ type Project struct {
 	UpdatedTime time.Time `bson:"updatedTime,omitempty" json:"updatedTime,omitempty" description:"updated time of the project"`
 }
 
+// PipelineRecord is all information about some pipeline building history.
+type PipelineRecord struct {
+	ID          string       `bson:"_id,omitempty" json:"id,omitempty" description:"id of the pipelineRecord"`
```
Comment (supereagle) [inline pkg/api/types.go]: s/pipelineRecord/pipeline record/g

File: pkg/api/types.go [hunk 2]
```diff
@@ -28,6 +28,53 @@ type Project struct {
...
+	PipelineID  string       `bson:"pipelineID,omitempty" json:"pipelineID,omitempty" description:"id of the related pipeline which the pipelineRecord belongs to"`
+	VersionID   string       `bson:"versionID,omitempty" json:"versionID,omitempty" description:"id of the related version which the pipelineRecord belongs to"`
+	Trigger     string       `bson:"trigger,omitempty" json:"trigger,omitempty" description:"trigger of the pipelineRecord"`
+	StageStatus *StageStatus `bson:"stageStatus,omitempty" json:"stageStatus,omitempty" description:"status of the latest stage"`
```
Comment (supereagle) [inline pkg/api/types.go]: Change to: status of each pipeline stage

File: pkg/api/types.go [hunk 3]
```diff
@@ -28,6 +28,53 @@ type Project struct {
...
+	// Failure means Cyclone had run with this stage and result in failure.
+	Failure Status = "Failed"
+	// Abort means the stage was aborted by some reason, and we can get the reason from the log.
+	Abort Status = "Aborted"
+)
+
+// TODO The status of every stage may be different.
+// StageStatus shows the information of every stage.
+type StageStatus struct {
+	CodeCheckout    *GeneralStageStatus `bson:"codeCheckout,omitempty" json:"codeCheckout,omitempty" description:"code checkout stage"`
```
Comment (supereagle) [inline pkg/api/types.go]: Change to: status of code checkout stage

File: pkg/api/types.go [hunk 4]
```diff
@@ -28,6 +28,53 @@ type Project struct {
...
+	Status      Status       `bson:"status,omitempty" json:"status,omitempty" description:"status of the pipelineRecord"`
+	StartTime   time.Time    `bson:"startTime,omitempty" json:"startTime,omitempty" description:"startTime of the pipelineRecord"`
+	EndTime     time.Time    `bson:"endTime,omitempty" json:"endTime,omitempty" description:"endTime of the pipelineRecord"`
+}
+
+// Status can be the status of some pipelineRecord or some stage
+type Status string
+
+const (
+	// Pending means Cyclone is preparing to run this stage.
```
Comment (supereagle) [inline pkg/api/types.go]: Pending represents the status that is triggered but still not running.

File: store/store.go [hunk 5]
```diff
@@ -60,9 +62,10 @@ func Init(s *mgo.Session) {
 func NewStore() *DataStore {
 	s := session.Copy()
 	return &DataStore{
-		s:                  s,
-		projectCollection:  s.DB(defaultDBName).C(projectCollectionName),
-		pipelineCollection: s.DB(defaultDBName).C(pipelineCollectionName),
+		s:                        s,
+		projectCollection:        s.DB(defaultDBName).C(projectCollectionName),
+    pipelineCollection:       s.DB(defaultDBName).C(pipelineCollectionName),
```
Comment (supereagle) [inline store/store.go]: Please go format this file.

File: pkg/api/types.go [hunk 6]
```diff
@@ -28,6 +28,53 @@ type Project struct {
...
+// PipelineRecord is all information about some pipeline building history.
+type PipelineRecord struct {
+	ID          string       `bson:"_id,omitempty" json:"id,omitempty" description:"id of the pipelineRecord"`
+	PipelineID  string       `bson:"pipelineID,omitempty" json:"pipelineID,omitempty" description:"id of the related pipeline which the pipelineRecord belongs to"`
+	VersionID   string       `bson:"versionID,omitempty" json:"versionID,omitempty" description:"id of the related version which the pipelineRecord belongs to"`
+	Trigger     string       `bson:"trigger,omitempty" json:"trigger,omitempty" description:"trigger of the pipelineRecord"`
+	StageStatus *StageStatus `bson:"stageStatus,omitempty" json:"stageStatus,omitempty" description:"status of the latest stage"`
+	Status      Status       `bson:"status,omitempty" json:"status,omitempty" description:"status of the pipelineRecord"`
+	StartTime   time.Time    `bson:"startTime,omitempty" json:"startTime,omitempty" description:"startTime of the pipelineRecord"`
+	EndTime     time.Time    `bson:"endTime,omitempty" json:"endTime,omitempty" description:"endTime of the pipelineRecord"`
```
Comment (supereagle) [inline pkg/api/types.go]: Variable names should not be used in description and comments, it is better to change to `end time of the pipeline record`. Please fix the same error in other places at the same time.

File: pkg/api/types.go [hunk 7]
```diff
@@ -28,6 +28,53 @@ type Project struct {
...
+
+// Status can be the status of some pipelineRecord or some stage
+type Status string
+
+const (
+	// Pending means Cyclone is preparing to run this stage.
+	Pending Status = "Pending"
+	// Running means Cyclone is running with this stage.
+	Running Status = "Running"
+	// Success means Cyclone had run with this stage and result in success.
```
Comment (supereagle) [inline pkg/api/types.go]: Success represents the status that finished and succeeded.

File: pkg/api/types.go [hunk 8]
```diff
@@ -28,6 +28,53 @@ type Project struct {
...
+	EndTime     time.Time    `bson:"endTime,omitempty" json:"endTime,omitempty" description:"endTime of the pipelineRecord"`
+}
+
+// Status can be the status of some pipelineRecord or some stage
+type Status string
+
+const (
+	// Pending means Cyclone is preparing to run this stage.
+	Pending Status = "Pending"
+	// Running means Cyclone is running with this stage.
```
Comment (supereagle) [inline pkg/api/types.go]: Running represents the status that is running.

File: pkg/api/types.go [hunk 9]
```diff
@@ -28,6 +28,53 @@ type Project struct {
...
+	// Success means Cyclone had run with this stage and result in success.
+	Success Status = "Success"
+	// Failure means Cyclone had run with this stage and result in failure.
+	Failure Status = "Failed"
+	// Abort means the stage was aborted by some reason, and we can get the reason from the log.
+	Abort Status = "Aborted"
+)
+
+// TODO The status of every stage may be different.
+// StageStatus shows the information of every stage.
```
Comment (supereagle) [inline pkg/api/types.go]: // StageStatus represents the collections of status for all stages.

File: pkg/api/types.go [hunk 10]
```diff
@@ -28,6 +28,53 @@ type Project struct {
 	UpdatedTime time.Time `bson:"updatedTime,omitempty" json:"updatedTime,omitempty" description:"updated time of the project"`
 }
 
+// PipelineRecord is all information about some pipeline building history.
```
Comment (supereagle) [inline pkg/api/types.go]: Move definition of `PipelineRecord` after `Pipeline` as `PipelineRecord` depends on `Pipeline`.

**Comment by zz-bmj** [comment]
@supereagle  please have a review of this pr again, it had fixed with your suggestions~

File: store/pipelineRecord.go [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+	"github.com/caicloud/cyclone/pkg/api"
+	"gopkg.in/mgo.v2/bson"
+)
+
+// CreatePipelineRecord creates the pipelineRecord, returns the pipelineRecord created.
+func (d *DataStore) CreatePipelineRecord(pipelineRecord *api.PipelineRecord) (*api.PipelineRecord, error) {
+	pipelineRecord.ID = bson.NewObjectId().Hex()
+	pipelineRecord.StartTime = time.Now()
+
+	err := d.pipelineRecordCollection.Insert(pipelineRecord)
```
Comment (supereagle) [inline store/pipelineRecord.go]: No need to declare the variable `err`, change to: 
```
if err := d.pipelineRecordCollection.Insert(pipelineRecord); err != nil {
  return nil, err
}
```
Please check the same error in other places.

**Comment by supereagle** [comment]
LGTM

**Pull request merged**